ENA-3950

haseeb-gh opened the pull request:
- moved fastq to sam implementation from Frankenstein to this repository and renames it to EnaFastq2Sam.
- changed feeder/eater terminologies and replaced them with producer/consumer.

haseeb-gh committed:
ENA-3950

- moved fastq to sam implementation from Frankenstein to this repository and renames it to EnaFastq2Sam.
- changed feeder/eater terminologies and replaced them with producer/consumer.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/Fastq2Sam.java:
```diff
@@ -42,7 +42,7 @@ public Fastq2Sam(String sampleName, String inputFile, String outputFile, String
-     * @param tempDir - Path to the temporary directory to be used during the process.. Pass null to use system default.
+     * @param tempDir - Path to the temporary directory to be used during the process. Pass null to use system default.
      */
     public Fastq2Sam(String sampleName, String inputFile1, String inputFile2, String outputFile, String tempDir) {
```
Suggestion to move to package picardfastq2sam to match enafastq2sam.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/enafastq2sam/EnaFastq2Sam.java:
```diff
@@ -0,0 +1,217 @@
+
+    public static void main( String[] args ) {
+        final Params p = new Params();
+        JCommander jc = new JCommander( p );
```
Generic comment. We should use https://picocli.info/ instead of JCommander. E.g. Webin-CLI uses picocli. Picocli is simply much better.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/enafastq2sam/EnaFastq2Sam.java:
```diff
@@ -0,0 +1,217 @@
+        }
+    }
+    
+    public void create( Params p ) throws SecurityException, FileNotFoundException, DataProducerException,
```
Please convert checked exceptions into runtime exceptions and remove throws clauses.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/enafastq2sam/EnaFastq2Sam.java:
```diff
@@ -0,0 +1,217 @@
+            NoSuchMethodException, IOException, DataConsumerException {
+
+        if( null == p.files || p.files.size() < 1 )
+            throw new IllegalStateException( "No input files" );
```
Not the correct type of exception. Should be IllegalArgumentException or EnaFastq2SamException extends RuntimeException.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/enafastq2sam/EnaFastq2Sam.java:
```diff
@@ -0,0 +1,217 @@
+            throw new IllegalStateException( "No input files" );
+        
+        if( p.verbose ) {
+            System.out.println( "Files to process: " );
```
This is a generic comment. We should prefer logging to System.out. Setting logging for command line use is tricky (just want to show severity and message) but possible.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/enafastq2sam/EnaFastq2Sam.java:
```diff
@@ -0,0 +1,217 @@
+            type = ILLUMINA_SETTINGS.valueOf( p.quality_type );
+        else
+            type = ILLUMINA_SETTINGS.valueOf( QualityNormalizer.valueOf( p.quality_encoding ),
+                                              QualityConverter.valueOf( p.quality_scoring ) );
```
Remove. We should delegate quality encoding and scale detection and transformation to Picard.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/enafastq2sam/EnaFastq2Sam.java:
```diff
@@ -0,0 +1,217 @@
+            } else {
+                if( p.verbose )
+                    System.out.println( "Paired file/read expected" );
+            }
```
Simplify. Throw an exception if the files name are same. Better still do this earlier in the code.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/enafastq2sam/EnaFastq2Sam.java:
```diff
@@ -0,0 +1,217 @@
+            
+        } else {
+            throw new IllegalArgumentException("More than two input files not allowed.");
+        }
```
Simplify. Test that we have 1 of 2 fastq files in the same place we do:

```
 if( null == p.files || p.files.size() < 1 )
            throw new IllegalStateException( "No input files" );
```

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/enafastq2sam/EnaFastq2Sam.java:
```diff
@@ -0,0 +1,217 @@
+        
+        ArrayList<AbstractDataProducer<?>> producers = new ArrayList<AbstractDataProducer<?>>();
+        
+        final QualityNormalizer normalizer = type.getNormalizer();
```
Remove. We should delegate quality normalization to Picard.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/enafastq2sam/EnaFastq2Sam.java:
```diff
@@ -0,0 +1,217 @@
+        
+        boolean again = false;
+        do {
+            for( AbstractDataProducer<?> feeder : producers ) {
```
Change feeder to producer.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/feeder/DataProducerException.java:
```diff
@@ -10,45 +10,40 @@
 */
 package uk.ac.ebi.ena.readtools.loader.common.feeder;
```
Package name should not be feeder.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/feeder/DataProducerException.java:
```diff
@@ -10,45 +10,40 @@
-public class 
-DataFeederException extends Exception
+public class
+DataProducerException extends Exception
```
Extend RuntimeException or EnaFastq2SamException extends RuntimeException. Don't use checked exceptions. I don't think we need ThreadName as part of the message.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/feeder/DataProducerEOFException.java:
```diff
@@ -10,38 +10,34 @@
-public class 
-DataFeederEOFException extends DataFeederException
+public class
+DataProducerEOFException extends DataProducerException
```
I don't think we need ThreadName as part of the message.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/feeder/DataProducerPanicException.java:
```diff
@@ -10,39 +10,35 @@
+DataProducerPanicException extends DataProducerException
 {
     private static final long serialVersionUID = 1L;
     private final String thread_name;
```
I don't think we need ThreadName as part of the message.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/feeder/DataProducer.java:
```diff
@@ -10,14 +10,14 @@
 {
-    public T feed() throws DataFeederEOFException, DataFeederException, DataFeederPanicException;
-    public DataFeeder<T> setEater( DataEater<T, ?> eater );
+    public T produce() throws DataProducerEOFException, DataProducerException, DataProducerPanicException;
```
Once changed to RunTimeExceptions remove throws clause.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/IlluminaSingleDataConsumer.java:
```diff
@@ -10,29 +10,29 @@
-public class 
-IlluminaSingleDataEater implements DataEater<DataSpot, IlluminaSpot>
+public class
+IlluminaSingleDataConsumer implements DataConsumer<DataSpot, IlluminaSpot>
```
Rename to SingleFastqConsumer. We are not only processing Illumina Fastqs.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/eater/DataConsumerException.java:
```diff
@@ -10,26 +10,23 @@
-public class 
-DataEaterException extends Exception
+public class
+DataConsumerException extends Exception
```
Extend RuntimeException or EnaFastq2BamException extends RuntimeException.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/eater/NullDataEater.java:
```diff
@@ -13,13 +13,13 @@
 
 public class 
-NullDataEater<T1> implements DataEater<T1, Void>
+NullDataEater<T1> implements DataConsumer<T1, Void>
```
Try to remove NullDataEater. Used only by FastqScanner I think that overrides consume in any case.

raskoleinonen replied:
Please remove NullDataEater. Used only by FastqScanner I think that overrides consume in any case.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/eater/PrintDataEater.java:
```diff
@@ -13,13 +13,13 @@
 
 public class 
-PrintDataEater<T1, T2> implements DataEater<T1, T2>
+PrintDataEater<T1, T2> implements DataConsumer<T1, T2>
```
Please rename to PrintDataConsumer. Please change the package name (currently 'eater').

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/IlluminaPairedDataConsumer.java:
```diff
@@ -15,12 +15,12 @@
-public class 
-IlluminaPairedDataEater extends AbstractPagedDataEater<DataSpot, IlluminaSpot>
+public class
+IlluminaPairedDataConsumer extends AbstractPagedDataConsumer<DataSpot, IlluminaSpot>
```
Rename to PairedFastqConsumer. We are not only processing Illumina Fastqs.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/sampler/SamplingDataEater.java:
```diff
@@ -13,12 +13,12 @@
 
 public class 
-SamplingDataEater<T1 extends DataSpot, T2> implements DataEater<T1, T2>
+SamplingDataEater<T1 extends DataSpot, T2> implements DataConsumer<T1, T2>
```
Please remove if not used. Please rename to SamplingDataConsumer if used.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/enafastq2sam/IlluminaBamConsumer.java:
```diff
@@ -0,0 +1,190 @@
+/**
+ * Accepts Illumina spot data and writes them out to a BAM file.
+ */
+public class IlluminaBamConsumer implements DataConsumer<IlluminaSpot, Object> {
```
Rename to Fastq2BamConsumer. This is not only used for Illumina Fastqs.

haseeb-gh replied:
i think all these eaters/consumers have the word Illumina in them because they accept an object of IlluminaSpot class. I dont really know why is that class called so. Maybe because of the terminologies or format used inside it?

https://github.com/enasequence/readtools/blob/master/src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/IlluminaSpot.java

raskoleinonen replied:
The IlluminaSpot is used to bring together paired reads. Better name would be 'FastqRead'. Spot is an alternative name for read on a physical experiment plate where the read is computed from a visual spot that changes intensity over time.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/enafastq2sam/IlluminaBamConsumer.java:
```diff
@@ -0,0 +1,190 @@
+                rec.setReadPairedFlag(false);
+                writer.addAlignment(rec);
+            }
+        } catch (QualityNormalizer.QualityNormaizationException ex) {
+            isOk = false;
+            ex.printStackTrace();
+            throw new RuntimeException( "FATAL: unable to normalize quality: " + iSpot.quals, ex );
```
Is anything throwing QualityNormalizer.QualityNormaizationException exception?

haseeb-gh replied:
yes. please see the method createSamRecord() below.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/bam/BamEater.java:
```diff
@@ -15,14 +15,14 @@
 public class 
-BamEater extends AbstractPagedDataEater<BamSpot, IlluminaSpot> 
-         implements DataEater<BamSpot, IlluminaSpot>
+BamEater extends AbstractPagedDataConsumer<BamSpot, IlluminaSpot>
```
Remove BamEater. I don't think there is any use for it anymore and that it is only used in ../loader/bam/LoaderTest.java that should be removed as well.

raskoleinonen commented on src/test/java/uk/ac/ebi/ena/readtools/loader/bam/LoaderTest.java:
```diff
@@ -19,9 +19,9 @@
+import uk.ac.ebi.ena.readtools.loader.common.feeder.DataProducerException;
 import uk.ac.ebi.ena.readtools.loader.fastq.IlluminaSpot;
 
 public class 
```
Remove ../loader/bam/LoaderTest.java. I think it is only used to test BamEater that is not needed anymore.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/bam/BamFeeder.java:
```diff
@@ -24,19 +24,19 @@
 
 public class 
-BamFeeder extends Thread implements DataFeeder<BamSpot>
+BamFeeder extends Thread implements DataProducer<BamSpot>
```
Remove BamFeeder. I don't see any use for it anymore after we have removed BamEater and associated classes.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/bam/BamSpot.java:
```diff
@@ -17,14 +17,14 @@
+import uk.ac.ebi.ena.readtools.loader.common.feeder.ProducibleData;
 
 public class 
 BamSpot implements Serializable
```
Remove BamSpot. I don't see any use for it anymore after we have removed BamEater, BamFeeder and related classes.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/cg/reads/CompleteGenomicsPairedEater.java:
```diff
@@ -13,16 +13,16 @@
 
 public class 
-CompleteGenomicsPairedEater extends AbstractDataEater<CompleteGenomicsBase, CGSpot>
+CompleteGenomicsPairedEater extends AbstractDataConsumer<CompleteGenomicsBase, CGSpot>
```
Remove CompleteGenomicsPairedEater. It is not needed anymore.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/cg/reads/CompleteGenomicsMap3.java:
```diff
@@ -18,9 +18,9 @@
+import uk.ac.ebi.ena.readtools.loader.common.feeder.ProducibleDataChecker;
 
 public class 
 CompleteGenomicsMap3 extends CompleteGenomicsBase implements Serializable 
```
Remove CompleteGenomicsMap3. It is not needed after we remove CompleteGenomicsPairedEater and related classes.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/cg/reads/CompleteGenomicsRead.java:
```diff
@@ -18,9 +18,9 @@
+import uk.ac.ebi.ena.readtools.loader.common.feeder.ProducibleDataChecker;
 
 public class 
 CompleteGenomicsRead extends CompleteGenomicsBase implements Serializable 
```
Remove CompleteGenomicsRead. It is not needed after we remove CompleteGenomicsPairedEater and related classes.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/cg/reads/Loader.java:
```diff
@@ -18,11 +18,11 @@
+import uk.ac.ebi.ena.readtools.loader.common.feeder.DataProducerException;
 
 public class 
 Loader
```
Remove .../loader/cg/reads/Loader.java. It is not needed after we remove CompleteGenomicsPairedEater and related classes.

raskoleinonen commented on src/test/java/uk/ac/ebi/ena/readtools/loader/cg/reads/TestEater.java:
```diff
@@ -23,11 +23,11 @@
+import uk.ac.ebi.ena.readtools.loader.common.feeder.DataProducerException;
 
 public class 
 TestEater
```
Remove ../loader/cg/reads/TestEater. It is not needed after we remove CompleteGenomicsPairedEater and related classes.

raskoleinonen commented on src/test/java/uk/ac/ebi/ena/readtools/loader/cg/reads/TestMap.java:
```diff
@@ -21,8 +21,8 @@
+import uk.ac.ebi.ena.readtools.loader.common.feeder.DataProducerException;
 
 public class 
 TestMap
```
Remove ../loader/cg/reads/TestMap. It is not needed after we remove CompleteGenomicsPairedEater and related classes.

raskoleinonen commented on src/test/java/uk/ac/ebi/ena/readtools/loader/cg/reads/TestRead.java:
```diff
@@ -21,8 +21,8 @@
+import uk.ac.ebi.ena.readtools.loader.common.feeder.DataProducerException;
 
 public class 
 TestRead
```
Remove ../loader/cg/reads/TestRead. It is not needed after we remove CompleteGenomicsPairedEater and related classes.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/enafastq2sam/EnaFastq2Sam.java:
```diff
@@ -0,0 +1,217 @@
+                    quality_scoring );
+        }
+    }
+}
```
Would be better to use https://picocli.info/. I could help with this.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/enafastq2sam/ILLUMINA_SETTINGS.java:
```diff
@@ -0,0 +1,82 @@
+import uk.ac.ebi.ena.readtools.loader.common.QualityConverter;
+import uk.ac.ebi.ena.readtools.loader.common.QualityNormalizer;
+
+public enum ILLUMINA_SETTINGS {
```
We should try to remove this class and delete all quality score transformation functionality to Picard.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/eater/DataConsumer.java:
```diff
@@ -10,11 +10,9 @@
+public interface DataConsumer<T1, T2> {
+    void cascadeErrors() throws DataConsumerException;
+    void consume(T1 object ) throws DataConsumerException;
+    void setConsumer(DataConsumer<T2, ?> dataConsumer);
```
It is not clear what is the purpose of setConsumer. Why do we have setConsumer inside Consumer. Should this even be in the interface of does it have a more specific role in the subclasses.

haseeb-gh replied:
I believe that is due the way these eaters/consumers are designed to work. It is expected that a given consumer may or may not want to pass information on to another consumer. Gives the idea of hierarchical consumer layout.

This interface also enforces that a consumer that is being set here must accept T2 as input. Which is the type of the object this consumer will produce upon consuming object of type T1.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/feeder/ProducibleData.java:
```diff
@@ -18,7 +18,7 @@
 @Target( {ElementType.FIELD } )
 public @interface
-FeedableData
+ProducibleData
```
Remove ProducibleData. I don't understand why we would need it for anything. If we need it make it clear what it is used for.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/feeder/ProducibleDataChecker.java:
```diff
@@ -18,6 +18,6 @@
 @Target( {ElementType.METHOD } )
 public @interface
-FeedableDataChecker
+ProducibleDataChecker
```
Remove ProducibleDataChecker. I don't understand why we would need it for anything. If we need it make it clear what it is used for.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/DataSpot.java:
```diff
@@ -70,28 +70,28 @@
-    @FeedableDataChecker
+    @ProducibleDataChecker
     public void
-    checkFeed() throws QualityNormaizationException, DataFeederException
+    checkFeed() throws QualityNormaizationException, DataProducerException
     {
         if( !params.allow_empty )
         {
             if( null == bases || null == quals 
                 || 0 == bases.length() || 0 == quals.length() )
-                throw new DataFeederException( params.line_no, "Empty lines not allowed" );
+                throw new DataProducerException( params.line_no, "Empty lines not allowed" );
         }
         normailzer.normalize( quals );
     }
```
Check if checkfeed is used after removing CompleteGenomics. If not used then remove.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/DataSpot.java:
```diff
@@ -218,7 +218,7 @@
     /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
     public void 
-    feedBaseName( InputStream is ) throws IOException, DataFeederException
+    feedBaseName( InputStream is ) throws IOException, DataProducerException
```
Remove feedBaseName. I don't think it is used for anything.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/DataSpot.java:
```diff
@@ -238,26 +238,26 @@
     // get bases
     public void 
-    feedBases( InputStream is ) throws IOException, DataFeederException
+    feedBases( InputStream is ) throws IOException, DataProducerException
```
Remove feedBases. I don't think it is used for anything except in DataSpotTest:testInvalidBaseCharactersException.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/DataSpot.java:
```diff
@@ -280,17 +280,17 @@
     // get name of quality line
     public void 
-    feedQualName( InputStream is ) throws IOException, DataFeederException
+    feedQualName( InputStream is ) throws IOException, DataProducerException
```
Remove feedQualName. I don't think it is used for anything.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/DataSpot.java:
```diff
@@ -280,17 +280,17 @@
     
     public void 
-    feedQuals( InputStream is ) throws IOException, QualityNormaizationException, DataFeederException
+    feedQuals( InputStream is ) throws IOException, QualityNormaizationException, DataProducerException
```
Remove feedQuals. I don't think it is used for anything.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/DataSpot.java:
```diff
@@ -22,9 +22,9 @@
+import uk.ac.ebi.ena.readtools.loader.common.feeder.ProducibleDataChecker;
 
 public class 
 DataSpot implements Serializable
```
Please remove any unused data members after removing unused methods (see below).

haseeb-gh replied:
All the methods you pointed are important and are being called using Java Reflection. I have added documentation notes on the two annotations above. That will help understand how all of this works together.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/IlluminaIterativeEaterIterator.java:
```diff
@@ -23,15 +23,15 @@
 
 public class 
-IlluminaIterativeEaterIterator implements Iterator<IlluminaSpot>, DataEater<IlluminaSpot, Object>
+IlluminaIterativeEaterIterator implements Iterator<IlluminaSpot>, DataConsumer<IlluminaSpot, Object>
```
Remove IlluminaIterativeEaterIterator. I think this is only used in IlluminaIterativeEaterIteratorTest. Also remove IlluminaIterativeEaterIteratorTest.

haseeb-gh replied:
This class is being used inside IlluminaIterativeConsumer which itself is being used by FastqScanner. FastqScanner on the other hand is in use by ReadsValidator which, im quite certain, I have seen being used inside another repository. So i dont think its possible to remove this class completely.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/eater/AbstractDataConsumer.java:
```diff
@@ -18,11 +18,11 @@
-public abstract class 
-AbstractDataEater<T1, T2> implements DataEater<T1, T2>
+public abstract class
+AbstractDataConsumer<T1, T2> implements DataConsumer<T1, T2>
```
Remove unused setOk(boolean is_ok). T1 and T2 should be restricted to concrete types (e.g. marker interfaces). We should log rather than write to System.out.

haseeb-gh replied:
This class resides in a common package and is probably meant to be used by anything that considers it useful. A marker interface can be created but I dont think it would be of any help unless it is given a meaningful name and we know what is it useful for exactly. Concrete classes such as IlluminaSingleDataConsumer, IlluminaPairedDataConsumer do define the types when they inherit these abstract classes so i suppose thats good enough.

raskoleinonen replied:
We need some restriction types for the templates otherwise it is impossible for the occasional reader (like me) to figure out the expected relationship of T1 and T2 an the AbstractDataConsumer. AbstractDataConsumer is not a container class that can accept any Object. Because T1 and T2 are restricted we should be explicit about what they are restricted to in the class definion.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/eater/DataConsumer.java:
```diff
@@ -10,11 +10,9 @@
-    void    cascadeErrors() throws DataEaterException;
-    void    eat( T1 object ) throws DataEaterException;
-    void    setEater( DataEater<T2, ?> dataEater );
+public interface DataConsumer<T1, T2> {
```
The template parameters T1 and T2 need to be restricted to some actual allowed types. Let's at least create a marker interfaces so that we can better understand what can go in here.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/eater/AbstractPagedDataConsumer.java:
```diff
@@ -31,8 +31,8 @@
-public abstract class 
-AbstractPagedDataEater<T1, T2> extends AbstractDataEater<T1, T2>
+public abstract class
+AbstractPagedDataConsumer<T1, T2> extends AbstractDataConsumer<T1, T2>
```
T1 and T2 should be restricted to concrete types (e.g. marker interfaces). We should log rather than write to System.out.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/sampler/FastqFileReader.java:
```diff
@@ -48,21 +48,21 @@
     read( InputStream             is, 
           final QualityNormalizer normalizer,
-          final String stream_marker ) throws SecurityException, DataFeederException, NoSuchMethodException, IOException
+          final String stream_marker ) throws SecurityException, DataProducerException, NoSuchMethodException, IOException
```
Make exceptions unchecked and remove throws clause. Delegate quality normalisation to Picard.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/sampler/SplitReader.java:
```diff
@@ -18,8 +18,8 @@
+import uk.ac.ebi.ena.readtools.loader.common.feeder.DataProducerException;
 import uk.ac.ebi.ena.readtools.loader.fastq.DataSpot;
 import uk.ac.ebi.ena.readtools.loader.fastq.DataSpot.DataSpotParams;
 
```
Remove SplitReader. I don't see any use for it. Called only from uk.ac.ebi.ena.readtools.sampler.main().

haseeb-gh replied:
Are we sure that no other component is using uk.ac.ebi.ena.readtools.sampler.Main as either a dependency or an executable?

raskoleinonen replied:
Yes, if Eugene is not using it (need to double check with him).

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/sampler/StripesDataEater.java:
```diff
@@ -19,16 +19,16 @@
 
 public class 
-StripesDataEater<T1 extends DataSpot, T2> implements DataEater<T1, T2>
+StripesDataEater<T1 extends DataSpot, T2> implements DataConsumer<T1, T2>
```
Remove StripesDataEater. Used only by StripesDataEater that can also be removed.

haseeb-gh replied:
This is in use by SplitReader.

raskoleinonen reviewed (changes requested):
Two main themes.
1. I think there are many classes we can simply remove.
2. Can we delegate quality encoding detection/transformation/scaling to Picard.

haseeb-gh committed:
ENA-3950

- PR followup.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/ena/Fastq2BamConsumer.java:
```diff
@@ -0,0 +1,186 @@
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Accepts Illumina spot data and writes them out to a BAM file.
+ */
```
Very minor comment: 'Accepts Fastq data ...'

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/ena/Fastq2BamConsumer.java:
```diff
@@ -0,0 +1,186 @@
+    public void consume(IlluminaSpot iSpot) throws DataConsumerException {
+        try {
+            Matcher matcher = VALID_DNA_CHARSET_PATTERN.matcher(iSpot.bases);
+            if( !matcher.matches() )
+                handleInvalidDnaCharset(iSpot.bases, matcher);
```
Generic comment. Please use Google Java format that e.g. always requires parenthesis after if.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/ena/Fastq2Sam.java:
```diff
@@ -0,0 +1,204 @@
+                System.out.println( " " + f_name );
+        }
+
+        Fastq2BamConsumer illuminaSpotToBamConsumer = new Fastq2BamConsumer(
```
Variable name should be 'fastq...' instead of 'illumina...'. Applies to all similar cases.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/ena/Fastq2Sam.java:
```diff
@@ -0,0 +1,204 @@
+        }
+
+        Fastq2BamConsumer illuminaSpotToBamConsumer = new Fastq2BamConsumer(
+                determineQualityNormalizer(p.quality_type), p.sample_name, p.data_file);
```
We should autodetect the quality encoding using Picard/htsjdk.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/common/reads/normalizers/htsjdk/IlluminaQualityNormalizer.java:
```diff
@@ -0,0 +1,12 @@
+
+    @Override
+    public void normalize(byte[] qualities) {
+        SolexaQualityConverter.getSingleton().convertSolexa_1_3_QualityCharsToPhredBinary(qualities);
```
Looks ok. Just double checking. Notes to myself:

convertSolexa_1_3_QualityCharsToPhredBinary(byte[] solexaQuals)
-----------------------------
Casava 1.3 stores phred-scaled qualities, but non-standard because they have 64 added to them rather than the standard 33.

corresponds to I think:

https://en.wikipedia.org/wiki/FASTQ_format#Quality I would expect this class to be called SolexaQualityNormalizer
-----------------------------
Starting with Illumina 1.3 and before Illumina 1.8, the format encoded a Phred quality score from 0 to 62 using ASCII 64 to 126 (although in raw read data Phred scores from 0 to 40 only are expected).
and
Starting in Illumina 1.5 and before Illumina 1.8, the Phred scores 0 to 2 have a slightly different meaning...

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/common/reads/normalizers/htsjdk/SolexaQualityNormalizer.java:
```diff
@@ -0,0 +1,12 @@
+
+    @Override
+    public void normalize(byte[] qualities) {
+        SolexaQualityConverter.getSingleton().convertSolexaQualityCharsToPhredBinary(qualities);
```
Looks ok. Just double checking. Notes to myself:

convertSolexaQualityCharsToPhredChars(byte[] solexaQuals)
---------------------------
Decodes an array of solexa quality ASCII chars into Phred ASCII space.

corresponds to I think:
---------------------------

https://en.wikipedia.org/wiki/FASTQ_format#Quality

Solexa/Illumina 1.0 format can encode a Solexa/Illumina quality score from -5 to 62 using ASCII 59 to 126 (although in raw read data Solexa scores from -5 to 40 only are expected)

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/common/reads/normalizers/htsjdk/StandardQualityNormalizer.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Override
+    public void normalize(byte[] qualities) {
+        SAMUtils.fastqToPhred(qualities);
```
Looks ok. Just double checking.

static void | fastqToPhred(byte[] fastq)
---------------------------
Converts printable qualities in Sanger fastq format to binary phred scores.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/ena/Fastq2Sam.java:
```diff
@@ -0,0 +1,204 @@
+        System.out.println( "DONE" );
+    }
+
+    private QualityNormalizer determineQualityNormalizer(String qualityType) {
+        if( qualityType == null || qualityType.isEmpty())
+            throw new IllegalArgumentException("Quality type must be provided.");
+
+        switch (FastqQualityFormat.valueOf(qualityType))  {
+            case Standard:
+                return new StandardQualityNormalizer();
+            case Solexa:
+                return new SolexaQualityNormalizer();
+            case Illumina:
+                return new IlluminaQualityNormalizer();
+            default:
+                throw new IllegalArgumentException("Unexpected fastq quality format provided : " + qualityType);
+        }
+    }
```
We should autodetect the quality encoding using Picard/htsjdk.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/producer/ProducibleData.java:
```diff
@@ -8,17 +8,24 @@
 @Target( {ElementType.FIELD } )
 public @interface
-FeedableData
+ProducibleData
```
Is it possible to replace ProducibleData annotation with a an interface with the required methods?

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/producer/ProducibleDataChecker.java:
```diff
@@ -8,16 +8,20 @@
 @Target( {ElementType.METHOD } )
 public @interface
-FeedableDataChecker
+ProducibleDataChecker
```
Is it possible to replace ProducibleDataChecker annotation with a an interface with the required methods?

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/IlluminaIterativeConsumer.java:
```diff
@@ -15,8 +15,8 @@
-public class 
-IlluminaIterativeEater implements Iterable<IlluminaSpot>
+public class
+IlluminaIterativeConsumer implements Iterable<IlluminaSpot>
```
Rename to FastqIterativeConsumer please.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/IlluminaIterativeConsumerIterator.java:
```diff
@@ -23,15 +23,16 @@
-public class 
-IlluminaIterativeEaterIterator implements Iterator<IlluminaSpot>, DataEater<IlluminaSpot, Object>
+public class
+IlluminaIterativeConsumerIterator implements Iterator<IlluminaSpot>, DataConsumer<IlluminaSpot, DataConsumable>
```
Rename to FastqIterativeConsumerIterator please.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/IlluminaSpot.java:
```diff
@@ -11,9 +11,10 @@
-public class 
-IlluminaSpot
+public class
+IlluminaSpot implements DataConsumable
```
Rename IluminaSpot to FastqSpot please. Remove DataConsumable interface and replace with FastSpot.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/consumer/DataConsumable.java:
```diff
@@ -0,0 +1,4 @@
+package uk.ac.ebi.ena.readtools.loader.common.consumer;
+
+public interface DataConsumable {
```
Rename IluminaSpot to FastqSpot please. Remove DataConsumable interface and replace with FastSpot.

haseeb-gh replied:
Are suggesting this interface be renamed to 'FastSpot' or should it be removed instead?

raskoleinonen replied:
DataConsumable interface to be removed and wherever this interface is used currently we should use FastqSpot instead.

haseeb-gh replied:
This is the marker interface that I added after you suggestion on limiting the T1 and T2 generic types. FastqSpot is a class. Along with DataSpot class, these two are used by the producers and consumers to read and write fastq data.

Please see the DataConsumer.java file to see how this interface is being used. FastqSpot and DataSpot implement this interface to become eligible to be used with DataConsumer and DataProducer types.

raskoleinonen replied:
OK, got it and spend a bit more time reading the code. This is actually linked to the removal of the @ProducibleData @ProducibleDataChecker annotations. If we replace these with an interface then this interface should be common to FastqSpot and DataSpot and this is currently 'DataConsumable'. For clarity I suggest we rename 'DataConsumable' to 'Spot' and add the methods from @ProducibleData @ProducibleDataChecker into 'Spot'. Then we can trivially see what is common between FastqSpot and DataSpot (i.e. what is 'Spot'). It is not clear (yet) why we need two different classes FastqSpot and DataSpot instead of just one.

raskoleinonen reviewed (commented):
Main comments: (1) we should autodetect the quality type (Picard’s fastq2Sam is doing this somehow), (2) when illumina was converted to fastq we should change variable names as well accordingly, (3) Is it possible to replace ProducibleData annotation with a an interface with the required methods, (4) Is it possible to replace ProducibleDataChecker annotation with a an interface with the required methods, (5) Rename IluminaSpot to FastqSpot please and remove DataConsumable interface and replace with FastSpot, (6) use Google Java Format. Also please think about adding more tests especially for fastq quality score transformation.

haseeb-gh committed:
ENA-3950

- PR followup.

haseeb-gh commented on src/main/java/uk/ac/ebi/ena/readtools/fastq/ena/Fastq2Sam.java:
```diff
@@ -0,0 +1,204 @@
+        }
+
+        FastqQualityFormat qualityFormat = Utils.detectFastqQualityFormat(p.files.get(0),
+                p.files.size() == 2 ? p.files.get(0) : null);
```
should be p.files.get(1) in the second argument.

raskoleinonen committed:
Minor changes. Open question about quality normalization and read index.

haseeb-gh committed:
ENA-3950

- fixed a small overlooked problem.

haseeb-gh committed:
Merge pull request #6 from enasequence/ENA-3950-rasko-1

Ena 3950 rasko 1

haseeb-gh committed:
Merge remote-tracking branch 'origin/ENA-3950' into ENA-3950

haseeb-gh committed:
ENA-3950

- major refactor of normalization and spot data reading code.

haseeb-gh committed:
ENA-3950

- refactored data spot reader to improve its readibility and reduce chances of bugs.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/utils/Utils.java:
```diff
@@ -15,6 +15,12 @@
+import uk.ac.ebi.ena.readtools.common.reads.normalizers.htsjdk.SolexaQualityNormalizer;
+import uk.ac.ebi.ena.readtools.common.reads.normalizers.htsjdk.StandardQualityNormalizer;
 
 import java.io.File;
```
Suggestion is to rename to QualityUtils.

haseeb-gh committed:
ENA-3950

- major refactor of FastqSpot class and all the relevant code.

raskoleinonen committed:
Minor changes and one validation message related bug fix.

haseeb-gh committed:
Merge pull request #7 from enasequence/ENA-3950-rasko-2

Minor changes and one validation message related bug fix.
